File: src/main/java/ravex/modules/player/AirPlace.java
package ravex.modules.player;

import ravex.modules.Category;
import ravex.modules.Module;
import ravex.parameter.BooleanParameter;
import ravex.utility.render.animate.SlideAnimation;
import ravex.utility.render.animate.FadeAnimation;
import ravex.utility.render.animate.SizeAnimation;
import ravex.utility.render.animate.BounceAnimation;
import net.minecraft.client.Minecraft;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.phys.Vec3;

public class AirPlace extends Module {
    public static final AirPlace INSTANCE = new AirPlace();
    
    // Static animation states for MixinLevelRenderer to query
    public static Vec3 highlightPos = null;
    public static float renderAlpha = 0.0f;
    public static double renderSize = 0.0;
    public static double renderYOffset = 0.0;

    // Render configuration parameter
    public final BooleanParameter render = new BooleanParameter("Render", true);

    // Animation utilities instances
    private final SlideAnimation slideAnim = new SlideAnimation();
    private final FadeAnimation fadeAnim = new FadeAnimation();
    private final SizeAnimation sizeAnim = new SizeAnimation();
    private final BounceAnimation bounceAnim = new BounceAnimation();

    private long lastPlaceTime = 0;

    private AirPlace() {
        super("AirPlace", Category.PLAYER);
        addParameter(render);
    }

    @Override
    protected void onEnable() {
        highlightPos = null;
        renderAlpha = 0.0f;
        renderSize = 0.0;
        renderYOffset = 0.0;
        slideAnim.reset();
        fadeAnim.reset();
        sizeAnim.reset();
        bounceAnim.reset();
    }

    @Override
    protected void onDisable() {
        highlightPos = null;
    }

    @Override
    public void onTick() {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null || mc.level == null) {
            highlightPos = null;
            return;
        }

        // 1. Check if holding a block in either hand and select the appropriate hand
        ItemStack main = mc.player.getItemInHand(InteractionHand.MAIN_HAND);
        ItemStack off = mc.player.getItemInHand(InteractionHand.OFF_HAND);
        boolean mainHolding = !main.isEmpty() && main.getItem() instanceof net.minecraft.world.item.BlockItem;
        boolean offHolding = !off.isEmpty() && off.getItem() instanceof net.minecraft.world.item.BlockItem;
        InteractionHand hand = mainHolding ? InteractionHand.MAIN_HAND : (offHolding ? InteractionHand.OFF_HAND : null);

        if (hand == null) {
            // Smoothly fade out when not holding blocks
            renderAlpha = fadeAnim.update(false, 0.15f);
            renderSize = sizeAnim.update(false, 0.15f);
            if (renderAlpha <= 0.01f) {
                highlightPos = null;
            }
            return;
        }

        // 2. Perform pick raycasting to find target block position in front of player
        double dist = 4.5;
        net.minecraft.world.phys.HitResult hit = mc.player.pick(dist, 1.0F, false);
        BlockPos targetPos;
        if (hit != null && hit.getType() == net.minecraft.world.phys.HitResult.Type.BLOCK) {
            targetPos = ((net.minecraft.world.phys.BlockHitResult) hit).getBlockPos();
        } else {
            Vec3 eye = mc.player.getEyePosition(1.0F);
            Vec3 look = mc.player.getViewVector(1.0F);
            Vec3 target = eye.add(look.x * dist, look.y * dist, look.z * dist);
            targetPos = BlockPos.containing(target);
        }

        // 3. Update animations smoothly
        if (render.getValue()) {
            renderAlpha = fadeAnim.update(true, 0.15f);
            renderSize = sizeAnim.update(true, 0.15f);
            renderYOffset = bounceAnim.update(0.05, 0.04);
            highlightPos = slideAnim.update(targetPos.getX(), targetPos.getY() + renderYOffset, targetPos.getZ(), 0.25);
        } else {
            // Instantly hide
            highlightPos = null;
            renderAlpha = 0.0f;
            renderSize = 0.0;
        }

        // 4. Place blocks if right-click key is pressed down, with a 200ms safe interval
        if (mc.options.keyUse.isDown()) {
            long now = System.currentTimeMillis();
            if (now - lastPlaceTime > 200) {
                Vec3 hitVec = Vec3.atCenterOf(targetPos);
                net.minecraft.world.phys.BlockHitResult blockHit = new net.minecraft.world.phys.BlockHitResult(
                    hitVec, net.minecraft.core.Direction.UP, targetPos, false
                );
                mc.gameMode.useItemOn(mc.player, hand, blockHit);
                mc.player.swing(hand);
                lastPlaceTime = now;
            }
        }
    }
}

File: src/main/kotlin/volthack/modules/player/AirPlace.kt
package volthack.modules.player

import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.InteractionHand
import net.minecraft.world.item.BlockItem
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3
import volthack.event.EventBus
import volthack.event.Render3DEvent
import volthack.event.TickEvent
import volthack.setting.Category
import volthack.setting.Module
import volthack.util.render.Render3DUtil

object AirPlace : Module("AirPlace", "Allows placing blocks in the empty air and renders a preview box of where the block will go", Category.PLAYER) {
    private val range by float("Reach Range", 4.5f, 2.0f, 6.0f, 0.1f)
    private val previewColor by color("Color", 0xFF00FF88.toInt())

    private var targetPos: BlockPos? = null
    private var lastPlaceTime = 0L

    init {
        EventBus.listen<TickEvent> { onTick() }
        EventBus.listen<Render3DEvent> { onRender3D(it) }
    }

    private fun onTick() {
        if (!enabled) {
            targetPos = null
            return
        }

        val mc = Minecraft.getInstance()
        val player = mc.player ?: return
        val world = mc.level ?: return

        // Check if holding a placeable block in either hand
        val hand = when {
            player.mainHandItem.item is BlockItem -> InteractionHand.MAIN_HAND
            player.offhandItem.item is BlockItem -> InteractionHand.OFF_HAND
            else -> {
                targetPos = null
                return
            }
        }

        // Trace looking vector
        val eyePos = player.getEyePosition(1.0f)
        val lookVec = player.getViewVector(1.0f)
        val targetX = eyePos.x + lookVec.x * range
        val targetY = eyePos.y + lookVec.y * range
        val targetZ = eyePos.z + lookVec.z * range
        val pos = BlockPos(
            Math.floor(targetX).toInt(),
            Math.floor(targetY).toInt(),
            Math.floor(targetZ).toInt()
        )

        // Check if the slot can be placed (must be air/replaceable)
        val state = world.getBlockState(pos)
        if (state.isAir || state.canBeReplaced()) {
            targetPos = pos

            // Place when right-clicking
            if (mc.options.keyUse.isDown && System.currentTimeMillis() - lastPlaceTime > 150) {
                lastPlaceTime = System.currentTimeMillis()

                val hitResult = BlockHitResult(
                    Vec3(pos.x + 0.5, pos.y + 0.5, pos.z + 0.5),
                    Direction.UP,
                    pos,
                    false
                )
                mc.gameMode?.useItemOn(player, hand, hitResult)
                player.swing(hand)
            }
        } else {
            targetPos = null
        }
    }

    private fun onRender3D(event: Render3DEvent) {
        if (!enabled) return
        val pos = targetPos ?: return

        val col = previewColor
        val r = (col shr 16) and 0xFF
        val g = (col shr 8) and 0xFF
        val b = col and 0xFF
        val a = (col shr 24) and 0xFF

        Render3DUtil.drawBlockOutline(
            event.modelViewMatrix,
            pos.x.toDouble() + 0.5,
            pos.y.toDouble(),
            pos.z.toDouble() + 0.5,
            1.0f, 1.0f,
            r, g, b, if (a == 0) 255 else a,
            true,
            2.0f
        )
    }
}
